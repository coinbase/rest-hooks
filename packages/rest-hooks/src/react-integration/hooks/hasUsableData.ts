import { FetchShape } from '~/resource';

/** If the invalidIfStale option is set we suspend if resource has expired */
export default function hasUsableData(
  cacheReady: boolean,
  fetchShape: Pick<FetchShape<any>, 'options'>,
) {
  return !(
    (fetchShape.options && fetchShape.options.invalidIfStale) ||
    !cacheReady
  );
}
