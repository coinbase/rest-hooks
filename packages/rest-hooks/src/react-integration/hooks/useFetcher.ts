import {
  FetchShape,
  DeleteShape,
  Schema,
  SchemaFromShape,
  ParamsFromShape,
  BodyFromShape,
} from 'rest-hooks/resource';
import { DispatchContext } from 'rest-hooks/react-integration/context';
import createFetch, {
  OptimisticUpdateParams,
} from 'rest-hooks/state/actions/createFetch';

import { useContext, useRef, useCallback } from 'react';

/** Build an imperative dispatcher to issue network requests. */
export default function useFetcher<
  Shape extends FetchShape<
    Schema,
    Readonly<object>,
    Readonly<object | string> | void
  >
>(
  fetchShape: Shape,
  throttle = false,
): Shape extends DeleteShape<any, any, any>
  ? (
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
    ) => ReturnType<typeof fetchShape['fetch']>
  : <
      UpdateParams extends OptimisticUpdateParams<
        SchemaFromShape<Shape>,
        FetchShape<any, any, any>
      >[]
    >(
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
      updateParams?: UpdateParams | undefined,
    ) => ReturnType<typeof fetchShape['fetch']> {
  const dispatch = useContext(DispatchContext);

  // we just want the current values when we dispatch, so
  // box the shape in a ref to make react-hooks/exhaustive-deps happy
  const shapeRef = useRef(fetchShape);
  shapeRef.current = fetchShape;

  const fetchDispatcher = useCallback(
    (
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
      updateParams?:
        | OptimisticUpdateParams<
            SchemaFromShape<Shape>,
            FetchShape<any, any, any>
          >[]
        | undefined,
    ) => {
      const action = createFetch(
        shapeRef.current,
        params,
        body,
        throttle,
        updateParams,
      );
      dispatch(action);
      return action.meta.promise;
    },
    [dispatch, throttle],
  );
  // any is due to the ternary that we don't want to deal with in our implementation
  return fetchDispatcher as any;
}
