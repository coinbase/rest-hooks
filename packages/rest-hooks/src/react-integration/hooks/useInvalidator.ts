import { useContext, useCallback, useRef } from 'react';
import { ReadShape, Schema } from 'rest-hooks/resource';
import { DispatchContext } from 'rest-hooks/react-integration/context';
import { INVALIDATE_TYPE } from 'rest-hooks/actionTypes';

/** Invalidate a certain item within the cache */
export default function useInvalidator<
  Params extends Readonly<object>,
  S extends Schema
>(fetchShape: ReadShape<S, Params>): (params: Params | null) => void {
  const dispatch = useContext(DispatchContext);
  const getFetchKeyRef = useRef(fetchShape.getFetchKey);
  getFetchKeyRef.current = fetchShape.getFetchKey;

  const invalidateDispatcher = useCallback(
    (params: Params | null) => {
      if (!params) return;
      dispatch({
        type: INVALIDATE_TYPE,
        meta: {
          url: getFetchKeyRef.current(params),
        },
      });
    },
    [dispatch],
  );

  return invalidateDispatcher;
}
