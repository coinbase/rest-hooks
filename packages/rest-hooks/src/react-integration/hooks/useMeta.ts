import { FetchShape } from 'rest-hooks/resource';
import { StateContext } from 'rest-hooks/react-integration/context';
import { selectMeta } from 'rest-hooks/state/selectors';
import { useContext, useMemo } from 'react';

/** Gets meta for a fetch key. */
export default function useMeta<Params extends Readonly<object>>(
  { getFetchKey }: Pick<FetchShape<any, any, Params>, 'getFetchKey'>,
  params: Params | null,
) {
  const state = useContext(StateContext);
  const key = params ? getFetchKey(params) : '';

  return useMemo(() => {
    if (!key) return null;
    return selectMeta(state, key);
  }, [key, state]);
}
