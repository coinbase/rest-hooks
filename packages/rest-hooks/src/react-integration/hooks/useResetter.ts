import { useContext, useCallback } from 'react';

import { DispatchContext } from '~/react-integration/context';
import { RESET_TYPE } from '~/types';

/** Returns a function to completely clear the cache of all entries */
export default function useResetter(): () => void {
  const dispatch = useContext(DispatchContext);

  const resetDispatcher = useCallback(() => {
    dispatch({
      type: RESET_TYPE,
    });
  }, [dispatch]);

  return resetDispatcher;
}
