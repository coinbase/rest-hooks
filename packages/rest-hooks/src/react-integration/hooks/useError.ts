import { ReadShape, Schema } from '~/resource';
import useMeta from './useMeta';

type UseErrorReturn<P> = P extends null ? undefined : Error;

/** Access a resource or error if failed to get it */
export default function useError<
  Params extends Readonly<object>,
  S extends Schema
>(
  fetchShape: ReadShape<S, Params>,
  params: Params | null,
  cacheReady: boolean,
): UseErrorReturn<typeof params> {
  const meta = useMeta(fetchShape, params);
  if (!params) return;
  if (!cacheReady) {
    if (!meta) return;
    if (!meta.error) {
      // this means we probably deleted the entity found in this result
      const err: any = new Error(
        `Resource not found in cache ${
          params ? fetchShape.getFetchKey(params) : ''
        }`,
      );
      err.status = 404;
      return err;
    } else {
      return meta.error as any;
    }
  }
}
