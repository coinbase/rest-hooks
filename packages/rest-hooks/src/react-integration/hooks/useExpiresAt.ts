import { ReadShape } from 'rest-hooks/resource';

import useMeta from './useMeta';

/** Returns whether the data at this url is fresh or stale */
export default function useExpiresAt<Params extends Readonly<object>>(
  fetchShape: Pick<ReadShape<any, Params>, 'getFetchKey'>,
  params: Params | null,
): number {
  const meta = useMeta(fetchShape, params);
  if (!meta) {
    return 0;
  }
  return meta.expiresAt;
}
