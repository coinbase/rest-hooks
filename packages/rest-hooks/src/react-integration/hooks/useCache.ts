import { StateContext } from 'rest-hooks/react-integration/context';
import { ReadShape, ParamsFromShape } from 'rest-hooks/resource';
import { useDenormalized } from 'rest-hooks/state/selectors';

import { useContext, useMemo } from 'react';

import useExpiresAt from './useExpiresAt';

/** Access a resource if it is available. */
export default function useCache<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'schema'>
>(fetchShape: Shape, params: ParamsFromShape<Shape> | null) {
  const expiresAt = useExpiresAt(fetchShape, params);
  // This computation reflects the behavior of useResource/useRetrive
  // It only changes the value when expiry or params change.
  // This way, random unrelated re-renders don't cause the concept of expiry
  // to change
  const expired = useMemo(() => {
    if (Date.now() <= expiresAt || !params) return false;
    return true;
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, params && fetchShape.getFetchKey(params)]);

  const state = useContext(StateContext);
  return useDenormalized(fetchShape, params, state, expired)[0];
}
