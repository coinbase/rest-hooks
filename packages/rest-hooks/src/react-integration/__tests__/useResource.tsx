import React, { Suspense } from 'react';
import { render } from '@testing-library/react';
import nock from 'nock';
import {
  CoolerArticleResource,
  UserResource,
  InvalidIfStaleArticleResource,
} from '__tests__/common';

// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderRestHook,
  makeCacheProvider,
  mockInitialState,
} from '../../../../test';
import { normalize } from '../../resource';
import { DispatchContext, StateContext } from '../context';
import { useResource } from '../hooks';
import { payload, users, nested } from './fixtures';

import { State } from '~/types';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <DispatchContext.Provider value={dispatch}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </DispatchContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalled();
  expect(dispatch.mock.calls.length).toBe(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function ArticleComponentTester({ invalidIfStale = false }) {
  const resource = invalidIfStale
    ? InvalidIfStaleArticleResource
    : CoolerArticleResource;
  const article = useResource(resource.detailShape(), {
    id: payload.id,
  });
  return (
    <div>
      <h3>{article.title}</h3>
      <p>{article.content}</p>
    </div>
  );
}

describe('useResource()', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  const fbmock = jest.fn();

  function Fallback() {
    fbmock();
    return null;
  }

  beforeAll(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload)
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '')
      .delete(`/article/${payload.id}`)
      .reply(200, {})
      .get(`/article-cooler/0`)
      .reply(403, {})
      .get(`/article-cooler/666`)
      .reply(200, '')
      .get(`/article-cooler/`)
      .reply(200, nested)
      .get(`/user/`)
      .reply(200, users);
  });

  afterAll(() => {
    nock.cleanAll();
  });

  beforeEach(() => {
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });
  afterEach(() => {
    renderRestHook.cleanup();
  });

  it('should dispatch an action that fetches', async () => {
    await testDispatchFetch(ArticleComponentTester, [payload]);
  });

  it('should dispatch fetch when sent multiple arguments', async () => {
    function MultiResourceTester() {
      const [article, user] = useResource(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
      return null;
    }
    await testDispatchFetch(MultiResourceTester, [payload, users]);
  });
  it('should throw same promise until both resolve', async () => {
    const renderRestHook = makeRenderRestHook(makeCacheProvider);
    jest.useFakeTimers();
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .delay(1000)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/user/`)
      .delay(2000)
      .reply(200, users);

    function MultiResourceTester() {
      try {
        const [article, user] = useResource(
          [
            CoolerArticleResource.detailShape(),
            {
              id: payload.id,
            },
          ],
          [UserResource.listShape(), {}],
        );
        return article;
      } catch (e) {
        // TODO: we're not handling suspense properly so react complains
        // When upgrading test util we should be able to fix this as we'll suspense ourselves.
        if (typeof e.then === 'function') {
          return e;
        } else {
          throw e;
        }
      }
    }
    const { rerender, result } = renderRestHook(MultiResourceTester);
    const firstPromise = result.current;
    expect(firstPromise).toBeDefined();
    expect(typeof firstPromise.then).toBe('function');

    jest.advanceTimersByTime(50);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(1000);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(2000);
    rerender();
    expect(result.current).toBe(firstPromise);

    // TODO: we're not handling suspense properly so react complains
    // When upgrading test util we should be able to fix this as we'll suspense ourselves.
    const oldError = console.error;
    console.error = () => {};
    jest.runAllTimers();
    await result.current;
    rerender();
    expect(result.current).toMatchInlineSnapshot(`
      CoolerArticleResource {
        "author": null,
        "content": "whatever",
        "id": 5,
        "tags": Array [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      }
    `);
    // eslint-disable-next-line require-atomic-updates
    console.error = oldError;
  });
  it('should NOT suspend if result already in cache and options.invalidIfStale is false', () => {
    const state: State<unknown> = mockInitialState([
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ]) as any;

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend even when result is stale and options.invalidIfStale is false', () => {
    const { entities, result } = normalize(
      payload,
      CoolerArticleResource.getEntitySchema(),
    );
    const fetchKey = CoolerArticleResource.detailShape().getFetchKey(payload);
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      indexes: {},
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => Promise.resolve()}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester />
          </Suspense>{' '}
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend if result is not stale and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      indexes: {},
      meta: {
        [fetchKey]: {
          date: Infinity,
          expiresAt: Infinity,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester invalidIfStale />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should suspend if result stale in cache and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      indexes: {},
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => Promise.resolve()}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester invalidIfStale />
          </Suspense>
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    render(tree);
    expect(fbmock).toHaveBeenCalled();
  });

  // taken from integration
  it('useResource() should throw errors on bad network', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(CoolerArticleResource.detailShape(), {
        title: '0',
      });
    });
    expect(result.current).toBe(null);
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('useResource() should throw errors on bad network (multiarg)', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource([
        CoolerArticleResource.detailShape(),
        {
          title: '0',
        },
      ]);
    });
    expect(result.current).toBe(null);
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('should resolve parallel useResource() request', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResource(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
    });
    await waitForNextUpdate();
    const [article, users] = result.current;
    expect(article instanceof CoolerArticleResource).toBe(true);
    expect(article.title).toBe(payload.title);
    expect(users).toBeDefined();
    expect(users.length).toBeTruthy();
    expect(users[0] instanceof UserResource).toBe(true);
  });

  it('should not suspend with no params to useResource()', () => {
    let article: any;
    const { result } = renderRestHook(() => {
      article = useResource(CoolerArticleResource.detailShape(), null);
      return 'done';
    });
    expect(result.current).toBe('done');
    expect(article).toBeUndefined();
  });
});
