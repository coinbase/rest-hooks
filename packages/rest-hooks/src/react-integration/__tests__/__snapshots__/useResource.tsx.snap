// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResource() should dispatch an action that fetches 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": [Function],
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": [Function],
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResource() should dispatch fetch when sent multiple arguments 2`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": Array [
      [Function],
    ],
    "throttle": true,
    "url": "GET http://test.com/user/",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
